lib/vrf.c

```diff
@@ -994,25 +994,50 @@ const char *vrf_get_default_name(void)
 	return vrf_default_name;
 }
 
-int vrf_bind(vrf_id_t vrf_id, int fd, const char *name)
+int vrf_bind(vrf_id_t vrf_id, int fd, const char *ifname)
 {
 	int ret = 0;
 	struct interface *ifp;
+	struct vrf *vrf;
 
-	if (fd < 0 || name == NULL)
+	if (fd < 0)
 		return fd;
-	/* the device should exist
-	 * otherwise we should return
-	 * case ifname = vrf in netns mode => return
-	 */
-	ifp = if_lookup_by_name(name, vrf_id);
-	if (!ifp)
+
+	if (vrf_id == VRF_UNKNOWN)
 		return fd;
+
+	/* can't bind to a VRF that doesn't exist */
+	vrf = vrf_lookup_by_id(vrf_id);
+	if (!vrf_is_enabled(vrf))
+		return -1;
+
+	if (ifname && strcmp(ifname, vrf->name)) {
+		/* binding to a regular interface */
+
+		/* can't bind to an interface that doesn't exist */
+		ifp = if_lookup_by_name(ifname, vrf_id);
+		if (!ifp)
+			return -1;
```

Comment:
I have a problem with that if vrf_bind is called within vrf_socket.
vrf_socket is supposed to have the same params whatever the vrf backend.
the last parameter ifname should be used only when backend is not netns.

that is why here I would not return -1 if backend is netns.

Reply:
This code is not called if the ifname is not set, please, reread the code carefully.

Reply:
I think you may experience issues with calling this when using netns, because `vrf_socket` switches back to the default netns after opening a socket and calls `vrf_bind` from the default netns. This is a problem of `vrf_socket`, not `vrf_bind`.